extensions/webpush.md

```diff
@@ -0,0 +1,101 @@
+
+To address these limitations, various push notification mechanisms have been designed. This specification standardizes an extension for Web Push.
+
+```
+      ┌────────────┐              ┌────────────┐
+      │            │  Subscribe   │            │
+      │            ├─────────────►│            │
+      │ IRC client │              │ IRC server │
+      │            │              │            │
+      │            │              │            │
+      └────────────┘              └─────┬──────┘
+             ▲                          │
+             │                          │
+        Push │                          │Push
+notification │                          │notification
+             │       ┌──────────┐       │
+             │       │          │       │
+             └───────┤ Web Push │◄──────┘
+                     │  Server  │
+                     │          │
+                     └──────────┘
+```
+
+Web Push is defined in [RFC 8030], [RFC 8291] and [RFC 8292].
+
+Although Web Push has been designed for the Web, it can be used on other platforms as well. Web Push provides a vendor-neutral standard to send push notifications.
+
+## Implementation
+
+The `webpush` capability allows clients to subscribe to Web Push and receive
+notifications for messages of interest.
+
+Once a client has subscribed, the server will send push notifications for a
+server-defined subset of IRC messages. Each push notification MUST contain
+exactly one IRC message as the payload, without the final CRLF.
+
+The messages follow the same capabilities and the same `RPL_ISUPPORT` as when
+the client registered for Web Push notifications.
+
+## `WEBPUSH` Command
+
+A new `WEBPUSH` command is introduced. It has a case-insensitive subcommand:
+
+    WEBPUSH <subcommand> <params...>
+
+### `VAPIDPUBKEY` Subcommand
+
+The `VAPIDPUBKEY` subcommand allows clients to query the [Voluntary Application Server Identification (VAPID)][RFC 8292] public key of the server. This can be used to decrypt notifications upon reception.
+
+The client can query the VAPID public key by sending the command:
+
+    WEBPUSH VAPIDPUBKEY
+
+The server will reply with a [URL-safe base64-encoded][RFC 4648 section 5] public key usable with the Elliptic Curve Digital Signature Algorithm (ECDSA) over the P-256 curve.
+
+    WEBPUSH VAPIDPUBKEY <key>
+
+### `REGISTER` Subcommand
+
+The `REGISTER` subcommand creates a new Web Push subscription.
+
+    WEBPUSH REGISTER <endpoint> <keys>
+
+The `<endpoint>` is an URL pointing to a push server, which can be used to send push messages for this particular subscription.
+
+`<keys>` is a string encoded in the message-tag format. The values are [URL-safe base64-encoded][RFC 4648 section 5]. It MUST contain at least:
```

Comment:
This uses the URL-safe base64 encoding because that's what the [W3C API uses](https://w3c.github.io/push-api/#dom-pushsubscription-tojson). So this allows Web clients to just use the result of `subscription.toJSON().keys` as-is.

But maybe that isn't such a good idea. The W3C API also doesn't provide any function to encode an arbitrary byte buffer to URL-safe base64, so users need to roll out their own if we use it elsewhere.

Instead we could just use regular base64, just like SASL does. Users can call `subscription.getKey()`.